src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -956,27 +958,33 @@ private void BlockDownloaded(uint256 blockHash, Block block)
                 {
                     try
                     {
-                        chainedHeader = this.chainedHeaderTree.FindHeaderAndVerifyBlockIntegrity(block);
+                        chainedHeader = this.chainedHeaderTree.FindHeader(block);
                     }
                     catch (BlockDownloadedForMissingChainedHeaderException)
                     {
                         this.logger.LogTrace("(-)[CHAINED_HEADER_NOT_FOUND]");
                         return;
                     }
-                    catch (IntegrityValidationFailedException integrityException)
-                    {
-                        this.peerBanning.BanAndDisconnectPeer(integrityException.PeerEndPoint, integrityException.BanDurationSeconds, $"Integrity validation failed: {integrityException.Error.Message}");
-
-                        this.logger.LogTrace("(-)[INTEGRITY_VERIFICATION_FAILED]");
-                        return;
-                    }
                 }
                 else
                 {
                     this.logger.LogDebug("Block '{0}' failed to be delivered.", blockHash);
                 }
             }
 
+            if (block != null)
+            {
```

Comment:
this block of code should be right after finding the header

Reply:
You said outside the lock? Or were you referring to a different lock?

Reply:
Correct, outside of the peer lock. 

I mean when we can have header null there (return null instead of throwing) and return in that case like we do now we won't need to check if the block is null again. By that time we know that block and header are not null . 

So just remove that if basically so we have integrity outside of the lock and right after we find header and handle if it's null

Reply:
I'm not entirely sure I follow - if `block` is null then the `else` branch above gets executed, but the method does not return. Therefore coming out of the lock `block` could either be `null` or not `null`, and we need to check it again.

Reply:
Ah, yes you are right, we do need to check, sorry. 

Forgot about cases where we want to process block being null (when puller failed to deliver)